File: lib/src/main/java/co/infinum/goldfinger/Crypto.java
package co.infinum.goldfinger;

import android.support.annotation.Nullable;
import android.support.v4.hardware.fingerprint.FingerprintManagerCompat;
import android.util.Base64;

public interface Crypto {

    /**
     * Encrypt value with unlocked cryptoObject. Return null if encryption fails.
     */
    @Nullable
    String encrypt(FingerprintManagerCompat.CryptoObject cryptoObject, String value);

    /**
     * Decrypt value with unlocked cryptoObject. Return null if decryption fails.
     */
    @Nullable
    String decrypt(FingerprintManagerCompat.CryptoObject cryptoObject, String value);

    class Default implements Crypto {

        private final Logger logger;

        Default(Logger logger) {
            this.logger = logger;
        }

        @Nullable
        @Override
        public String encrypt(FingerprintManagerCompat.CryptoObject cryptoObject, String value) {
            try {
                byte[] encryptedBytes = cryptoObject.getCipher().doFinal(value.getBytes());
                return Base64.encodeToString(encryptedBytes, Base64.DEFAULT);
            } catch (Exception e) {
                logger.log(e);
                return null;
            }
        }

        @Nullable
        @Override
        public String decrypt(FingerprintManagerCompat.CryptoObject cryptoObject, String value) {
            try {
                byte[] decodedBytes = Base64.decode(value, Base64.DEFAULT);
                return new String(cryptoObject.getCipher().doFinal(decodedBytes));
            } catch (Exception e) {
                logger.log(e);
                return null;
            }
        }
    }
}

File: lib/src/main/kotlin/co.infinum.goldfinger/Crypto.kt
package co.infinum.goldfinger

import android.support.v4.hardware.fingerprint.FingerprintManagerCompat
import android.util.Base64
import javax.crypto.Cipher

interface Crypto {

    /**
     * Encrypt value with unlocked cryptoObject. Return null if encryption fails.
     */
    fun encrypt(cryptoObject: FingerprintManagerCompat.CryptoObject, value: String): String?

    /**
     * Decrypt value with unlocked cryptoObject. Return null if decryption fails.
     */
    fun decrypt(cryptoObject: FingerprintManagerCompat.CryptoObject, value: String): String?

    class Default(private val logger: Logger) : Crypto {

        override fun encrypt(cryptoObject: FingerprintManagerCompat.CryptoObject, value: String): String? {
            return try {
                val cipher = cryptoObject.cipher as Cipher
                val encryptedBytes = cipher.doFinal(value.toByteArray())
                Base64.encodeToString(encryptedBytes, Base64.DEFAULT)
            } catch (e: Exception) {
                logger.log(e)
                null
            }
        }

        override fun decrypt(cryptoObject: FingerprintManagerCompat.CryptoObject, value: String): String? {
            return try {
                val cipher = cryptoObject.cipher as Cipher
                val decryptedBytes = cipher.doFinal(Base64.decode(value, Base64.DEFAULT))
                String(decryptedBytes)
            } catch (e: Exception) {
                logger.log(e)
                null
            }
        }
    }
}